nixos/modules/services/x11/desktop-managers/elementary.nix

```diff
@@ -0,0 +1,212 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.services.xserver.desktopManager.elementary;
+
+  # Remove packages of ys from xs, based on their names
+  removePackagesByName = xs: ys:
+    let
+      pkgName = drv: (builtins.parseDrvName drv.name).name;
+      ysNames = map pkgName ys;
+    in
+      filter (x: !(builtins.elem (pkgName x) ysNames)) xs;
+
+  nixos-gsettings-desktop-schemas = pkgs.elementary.elementary-gsettings-schemas.override {
+    extraGSettingsOverridePackages = cfg.extraGSettingsOverridePackages;
+    extraGSettingsOverrides = cfg.extraGSettingsOverrides;
+  };
+
+in
+
+{
+  options = {
+
+    services.xserver.desktopManager.elementary = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Enable the elementary desktop manager";
+      };
+
+      sessionPath = mkOption {
+        default = [];
+        example = literalExample "[ pkgs.gnome3.gpaste ]";
+        description = ''
+          Additional list of packages to be added to the session search path.
+          Useful for GSettings-conditional autostart.
+
+          Note that this should be a last resort; patching the package is preferred (see GPaste).
+        '';
+        apply = list: list ++
+        [
+          pkgs.elementary.elementary-dpms-helper
+          pkgs.elementary.elementary-settings-daemon # TODO: I shouldn't need to do this
```

Comment:
Any ideas @jtojnar?

Reply:
You will need to add the package to

* `etc/xdg/autostart`: (I think) to [`XDG_CONFIG_DIRS`](https://specifications.freedesktop.org/autostart-spec/autostart-spec-latest.html#idm140434866702736) for the executable launched by the session desktop file; see Flashback for [an example](https://github.com/NixOS/nixpkgs/blob/869245c44822c21c1efca1ec46923f6615990342/pkgs/desktops/gnome-3/misc/gnome-flashback/default.nix#L143)
* `lib/udev/rules.d`: to `services.udev.packages`
* `share/gsettings-schemas`: to closure of packages that need those schemas
* `share/polkit-1/actions`: to `environment.systemPackages` (polkit looks to system profile when enabled)

Reply:
Same goes for the other things in `sessionPath`. And every item that remains should be probably be accompanied with a comment why it is there.

Reply:
Thanks for the information. Though my vague comment here certainly isn't much help to future me, as I actually don't know what I didn't do :smile:  

Only thing that sounds foreign to me is the point on `etc/xdg/autostart`. What does that have to do with the settings daemon? I've patched the autostarts.

Reply:
IIRC, `gnome-session` runs desktop files in `$XDG_CONFIG_DIRS/autostart`, therefore, if we want to have it ran, we need to add it there.